Documentation/guides/internals/debug-jni-objrefs.md

```diff
@@ -0,0 +1,400 @@
+/*    28 */    at Java.Lang.Object.Dispose() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 210
+/*    29 */    at Scratch.GrefCrashNet6.MainActivity.JniActivationCrash() in …/Scratch.GrefCrashNet6/Scratch.GrefCrashNet6/MainActivity.cs:line 35
+/*    30 */    at Scratch.GrefCrashNet6.MainActivity.OnCreate(Bundle savedInstanceState) in …/Scratch.GrefCrashNet6/Scratch.GrefCrashNet6/MainActivity.cs:line 16
+/*    31 */    at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-31/mcw/Android.App.Activity.cs:line 2781
+/*    32 */    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 121
+```
+
+Line 1: we're creating a JNI Global Reference for a `Java.Lang.Runnable` instance, with managed stack trace.
+
+Line 14 is how we know we're creating a `Java.Lang.Runnable` instance.  The `handle 0x1cda` value matches the `new-handle 0x1cda/G` value on Line 1.
+
+Line 16: we create a new JNI Global Reference with value 0x1cea.
+
+Line 23: we destroy the JNI Global Reference 0x1cda.
+
+This is where we need to "know things" and read carefully: line 28 has `Java.Lang.Object.Dispose()` in the stack trace.
+From this we must *infer* that the instance associated with JNI Global Reference 0x1cda has been disposed, which removes
+the association between *key_handle 0x466b26f* and the `Java.Lang.Runnable` *instance*.  This also tells us where
+`Java.Lang.Object.Dispose()` was called from: within `JniActivationCrash()`, in `MainActivity.cs:35`.
+
+[`logcat-parse`](#logcat-parse) can be used to find the peer via the `key_handle` value, and determine its state:
+
+```
+% mono ~/Developer/src/xamarin/Java.Interop/bin/Debug/logcat-parse.exe grefs.txt
+
+csharp> var r = grefs.AllocatedPeers.Where(p => p.KeyHandle=="0x466b26f");
+csharp> r;
+{ PeerInfo(State='Collected' JniType='mono/java/lang/Runnable' McwType='Java.Lang.Runnable' KeyHandle=0x466b26f Handles=1) }
+```
+
+The `PeerInfo.State` value of `Collected` means that the instance has been collected.
+
+### Dalvik Abort Crash Logs
+
+Consider the [crash via Dalvik Abort](#crash-dalvik-abort):
+
+```
+java_vm_ext.cc:579] JNI DETECTED ERROR IN APPLICATION: JNI ERROR (app bug): jobject is an invalid local reference: 0x75 (deleted reference at index 7 in a table of size 7)
+```
+
+Of note is that the abort message mentions *local*, in "an invalid local reference".  This is a sign that we need to collect the JNI Local Reference logs:
+
+```sh
+adb shell setprop debug.mono.log lref,gref
+
+# run the app…
+
+adb shell run-as @PACKAGE-NAME@ cat files/.__override__/grefs.txt > grefs.txt
+adb shell run-as @PACKAGE-NAME@ cat files/.__override__/lrefs.txt > lrefs.txt
```

Comment:
Perhaps explain where `@PACKAGE-NAME@` can be looked up? I wouldn't be surprised if many people wouldn't know what their package name actually is, it being hidden behind properties in the project file